### ISSUE TITLE
Add explainability to the connectivity analysis output

### ISSUE THREAD

**Issue by adisos** [np-guard/vpc-network-config-analyzer#42]
Add as a new command line query , input: src / dst (endpoint/subnet/)  , optionally protocol and port to explore 
output: explain which resources (SG/NACL) are relevant for connectivity blocked/allowed , and which rules in the resource.

if blocked: which NACLs is blocking (and if there is deny rule / deny by default) , and which SGs can allow (if exist)
if allowed: which NACLs + SGs allow (and which rules)

phase2: gateways / LBs / PGW (+ ip-address as src/dst )  / TGW

**Comment by ShiriMoran** [comment]
Stage 1: 
input:
src (dst) which is a network interface 
dst (src) which is an external IP
and assuming a single sg that governs the traffic from the given network interface
produce the following output:
1. Is the connection enabled or blocked
2. What are the rules in the sg enabling/blocking the connection or is it just by default with no specific rules addressing this connection

This may be considered as some type of a POC

**Comment by adisos** [comment]
Few comments about proposed output:
- (1) Is the connection enabled or blocked -- if enabled, I think it will also help to add the connection properties.
- (2) default setting - how do we know if a SG is the default one or a one that overrides the default SG? in the input config object contains the configured resources without such indication. (there is no situation where the config object has no SG and the analyzer assumes then a "default" resource to use).

**Comment by ShiriMoran** [comment]
https://github.com/np-guard/vpc-network-config-analyzer/issues/42#issuecomment-1853471622
@adisos I assume you refer to the output below.
1) currently only `SG` are handles, and connections are only enablers. Makes sense, I'll add it.
2) Agrees, but trying to have small `PR`s lets defer it to the next PR. The relevant check that we will need to add here is in a way independent in what we have already done.
```
There is a connection between vsi2-ky[10.240.20.4] and vsi3b-ky[10.240.30.4].
Egress Rules:
~~~~~~~~~~~~~
Security Groups Rules
---------------------
enabling rules from sg2-ky:
        index: 5, direction: outbound, protocol: all, cidr: 10.240.30.0/24
        index: 6, direction: outbound,  conns: protocol: tcp,  dstPorts: 1-65535, cidr: 10.240.20.4/32,10.240.30.4/32

Ingress Rules:
~~~~~~~~~~~~~~
Security Groups Rules
---------------------
enabling rules from sg2-ky:
        index: 7, direction: inbound,  conns: protocol: tcp,  dstPorts: 1-65535, cidr: 10.240.20.4/32,10.240.30.4/32

---------------------------------------------------------------------------------------------------------------------
There is a connection between vsi2-ky[10.240.20.4] and vsi1-ky[10.240.10.4].

Egress Rules:
~~~~~~~~~~~~~
Security Groups Rules
---------------------
enabling rules from sg2-ky:
        index: 1, direction: outbound, protocol: all, cidr: 10.240.10.0/24

Ingress Rules:
~~~~~~~~~~~~~~
Security Groups Rules
---------------------
enabling rules from sg1-ky:
        index: 3, direction: inbound, protocol: all, cidr: 10.240.20.4/32,10.240.30.4/32

```

**Comment by zivnevo** [comment]
Use cases for specifying connection details (protocol and port):
* "As a security person I would really like to know why TCP 22 is open from my laptop to VSI-5, as this is a security risk for my company. TCP 443 is also open between these two, but I'm totally OK with it, and couldn't care less about the rules involved."
* "As a developer I want to understand why VSI-3 fails getting answers for its DNS queries (UDP 53), while connecting to a known IP via TCP 8080 works."

**Comment by ShiriMoran** [comment]
[2nd PR](#266 ) includes list of ingress and egress rules relevant to given `src` to `dst` in case there is a connection or, in case there is no connection, information regarding the blocking - ingress, egress or both. 

Next PR: 
1. Input can be a `vsi` whether than a network interface in which case all relevant network interfaces will be considered. 
1. Add the existing connection in the output
1. If a specific enabling rule is the default one, mention it

The last two items are these mentioned by @adisos  https://github.com/np-guard/vpc-network-config-analyzer/issues/42#issuecomment-1853471622

**Comment by ShiriMoran** [comment]
1 is solved by PR #270

**Comment by ShiriMoran** [comment]
Regarding item 2 above:
Currently working on `SG` in `IBM VPC`, I got 3 different answers for the question: what are the default `SG` rules?
In the [documentation](https://cloud.ibm.com/docs/vpc?topic=vpc-using-security-groups&interface=ui) it says "by default, a security group is set up with rules that deny all inbound traffic and permit all outbound traffic. As new rules are added to a security group, the new rules redefine the scope of permitted inbound or outbound traffic."
In the [API spec](https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/swagger-ui.html?version=today#/Security%20groups/create_security_group) it says "If unspecified, no rules will be created, resulting in all traffic being denied."
Finally, when actually creating a VPC then the default SG automatically enables all outbound traffic and allows inbound traffic from the same SG.
Am I missing something? Any idea how to resolve this? Who to ask?

**Comment by ShiriMoran** [comment]
2 is solved by #286

**Comment by ShiriMoran** [comment]
This issue seems to be covered by the other explainability issues; specifically #387 and #368 covers V0.4
May I close it? @zivnevo @adisos
